### ISSUE TITLE
New mixed_float16 results by a very slow training

### ISSUE THREAD

**Issue by fitoule** [google/automl#351]
hi I train my custom dataset on a RTX NVidiaCard. I tested the new mixed_float16 : it consumes less memory on the GPU (half ) but finally the training is very very slow now 
about 1.8 steps/s with float32 by default
and 0.2 now with mixed_float16

?

**Comment by fitoule** [comment]
one good point is now  I can train D4 but it will take many days :(

**Comment by fitoule** [comment]
I tested on colab. It works well on it. mixed_float16 is faster. So it is an issue on NVIDIA only. Anybody knows why ?

**Comment by YangZhang4065** [comment]
Same observation here. Switching to mixed_float16 slows the same training code's steps-per-second from 1.6 to 0.2.

**Comment by fitoule** [comment]
I managed to use it on my NVIDIA RTX 2060 by adding this patch in  det_model_fn.py
  if mode == tf.estimator.ModeKeys.TRAIN:
    optimizer = tf.train.MomentumOptimizer(
        learning_rate, momentum=params['momentum'])
    #PATCH
    optimizer  = tf.train.experimental.enable_mixed_precision_graph_rewrite(optimizer) 	
    #update_ops = optimizer.minimize(total_loss) 
    #END PATCH
    if params['use_tpu']:
      optimizer = tf.tpu.CrossShardOptimizer(optimizer)

Try this but remove the hparam precision=mixed_float16 otherwise it will use Keras method that don't work let the precision by default

For me I can now train the D4 on my RTX !! ( with defaulkt size of 1024 )
and I go from 1.6 steps sec to 2.8  ;)

**Comment by staeff777** [comment]
I'll try it, fitoule.  Thanks.
So, it is just one line change for the optimizer.
You didn't change the update_ops ?

**Comment by staeff777** [comment]
Can you please state your hyperparams?

**Comment by fitoule** [comment]
for hparam i just set the num_classes.
I did'nt applied the minimize loss because It seems to be done later in the code. But I m not sure we have to test both.

Note that only RTX Nvidia Card minimum supports the mixed mode.

**Comment by staeff777** [comment]
I was required to not remove use_bfloat16=false (probably obvious). Afterwards I got an increase of 1.9 to 2.3 steps/sec.

**Comment by YangZhang4065** [comment]
Thank you @fitoule . The trick worked for me after upgrading my TensorFlow from 1.15 to 2.10. Obviously swish will trigger some error if someone tries to use mixed precision with TF1.15.

**Comment by fitoule** [comment]
I found a better option : this patch in det_model_fn.py

if mode == tf.estimator.ModeKeys.TRAIN:
    optimizer = tf.train.MomentumOptimizer(
        learning_rate, momentum=params['momentum'])
    #PATCH
    tf.train.experimental.enable_mixed_precision_graph_rewrite(
    	optimizer, loss_scale='dynamic'
	)
    #END PATCH
    if params['use_tpu']:

and same above do not use precision=mixed_float16 ( it is the keras method that works well on colab cards but not on NVIDIA RTX )

**Comment by fitoule** [comment]
guys I found another trick .... the last one  ... 
I upgraded my tensor flow and others package the same version than, colab that works fine. 
**tensorflow               2.2.0rc4**    

And .... It works with the implementation keras of the author and it 's even faster !!

Thank you we can close this issue

**Comment by fitoule** [comment]
for using Nvidia RTX mixed_float16 mode  : upgrade tensor flow to 2.2.0rc4

**Comment by elv-xuwen** [comment]
Thank you @fitoule for the trick. But somehow it doesn't work for me. I'm training D4 (batchsize==1) on GeForce RTX 2080 Ti with the trick but GPU Memory usage stays the same, both are around 10493MiB. Can you tell me your GPU Memory usage? Thank you so much!

Here's some part of my log:

> 2020-05-20 01:40:33.805078: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1917] Running auto_mixed_precision graph optimizer
> 2020-05-20 01:40:33.805285: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1298] No whitelist ops found, nothing to do
> 2020-05-20 01:40:42.565715: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1917] Running auto_mixed_precision graph optimizer
> 2020-05-20 01:40:42.978026: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1873] Converted 2946/37300 nodes to float16 precision using 397 cast(s) to float16 (excluding Const and Variable casts)
> 2020-05-20 01:41:23.931776: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcudnn.so.7
> 2020-05-20 01:41:26.521480: I tensorflow/stream_executor/platform/default/dso_loader.cc:44] Successfully opened dynamic library libcublas.so.10
> 2020-05-20 01:41:36.253530: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1917] Running auto_mixed_precision graph optimizer
> 2020-05-20 01:41:36.672607: I tensorflow/core/grappler/optimizers/auto_mixed_precision.cc:1873] Converted 2946/37297 nodes to float16 precision using 397 cast(s) to float16 (excluding Const and Variable casts)

**Comment by rex-yue-wu** [comment]
I encountered a very similar issue -- using `mixed_float16` results very slow training. However, the strange thing that I figured out is that where to invoke `mixed_float16` matters. In particular, the following two ways A and B, A runs much much slower than B, and A actually results 100% CPU usage on my machine. 

```python
# method A
training_dataset = create_tensorflow_dataset(...)
keras.mixed_precision.set_global_policy("mixed_float16")
model = create_keras_model(..)
model.fit(training_dataset, ...)
```

```python
# method B
keras.mixed_precision.set_global_policy("mixed_float16")
training_dataset = create_tensorflow_dataset(...)
model = create_keras_model(..)
model.fit(training_dataset, ...)
```

My guess is that the global policy somewhat impact the data generation process, and makes the batch preparation step much much longer than it should be.
